File: Behavioral/Lab23/Typescript/src/estate_subsystem.ts
import { AbstractSubsystem } from "./abstract_subsystem";
import { Agency } from "./Entities/agency";
import { AgencyObject } from "./Entities/agency_object";
import { RealEstate } from "./Entities/real_estate";

export class EstateSubsystem implements AbstractSubsystem {
    public MainAction(objects:Array<AgencyObject>, notify:boolean = true):void{
        let sb = "";
        const estates = objects as Array<RealEstate>;
        for (const estate of estates) {
            sb += `Estate ${estate.address}\n`;
        }
        if(notify)
            Agency.Notify("realtor");
        console.log(sb);
    }
}

File: Behavioral/Lab17/C#/EstateSubsystem.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Entities;

namespace C_
{
    public class EstateSubsystem : AbstractSubsystem
    {
        public override void MainAction(IEnumerable<AgencyObject> agencyObjects, bool notify = true)
        {
            var realEstates = agencyObjects as IEnumerable<RealEstate>;

            var sb = new StringBuilder();
            sb.Append($"Real Estates protfolio:\n");
            foreach (var estate in realEstates)
            {
                sb.Append($"Estate {estate.Address}\n");
            }
            if (notify)
                Agency.GetInstance().Notify("Realtors protfolio");
            Console.WriteLine(sb.ToString());
        }

    }
}
